[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows.

etaf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146481
* #147347

This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

etaf committed:
[Windows][Inductor] Enable Inductor UT on XPU Windows.

[ghstack-poisoned]

etaf committed:
Update on "[WIP][Windows][Inductor] Enable Inductor UT on XPU Windows."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

etaf committed:
Update on "[WIP][Windows][Inductor] Enable Inductor UT on XPU Windows."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/triton_bundler.py:
```diff
@@ -261,5 +262,7 @@ def read_and_emit(
                         # Just append one of them without the extension
                         kernel_names.append(Path(artifact.filename).stem)
                 # Atomic on POSIX systems
+                if os.path.exists(directory):
+                    shutil.rmtree(directory)
```
Why is this change needed?  Is there some sort of race happening?

The atomic comment above should apply to the os.replace.

cc @oulgen @jamesjwu

oulgen replied:
@jansel I assume this is a Windows problem since os.replace is only atomic on POSIX system :/

jansel replied:
Yeah, maybe we need to use FileLock on Windows.  This new change seems not thread safe, while the old one does.

etaf replied:
Thanks @jansel @oulgen .
This change is actually to fix the following case failure:
```
======================================================================
ERROR: test_cache_hot_load_device_xpu_bfloat16_dynamic_False (__main__.TestFxGraphCache)
Verify that we can populate and hot load functions from the cache.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\xinanlin\pytorch\torch\testing\_internal\common_utils.py", line 3155, in wrapper
    method(*args, **kwargs)
  File "c:\xinanlin\pytorch\torch\testing\_internal\common_utils.py", line 551, in instantiated_test
    test(self, **param_kwargs)
  File "C:\Users\sdp\miniforge3\envs\xinanlin\lib\contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "C:\Users\sdp\miniforge3\envs\xinanlin\lib\contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "C:\xinanlin\pytorch\test\inductor\test_codecache.py", line 361, in test_cache_hot_load
    compiled_result = compiled_fn(a, b)
  File "c:\xinanlin\pytorch\torch\_dynamo\eval_frame.py", line 589, in _fn
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
  File "c:\xinanlin\pytorch\torch\_dynamo\output_graph.py", line 1509, in _call_user_compiler
    raise BackendCompilerFailed(
  File "c:\xinanlin\pytorch\torch\_dynamo\output_graph.py", line 1488, in _call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
  File "c:\xinanlin\pytorch\torch\_dynamo\repro\after_dynamo.py", line 150, in __call__
    compiled_gm = compiler_fn(gm, example_inputs)
  File "c:\xinanlin\pytorch\torch\__init__.py", line 2339, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "c:\xinanlin\pytorch\torch\_inductor\compile_fx.py", line 2168, in compile_fx
    return aot_autograd(
  File "c:\xinanlin\pytorch\torch\_dynamo\backends\common.py", line 101, in __call__
    cg = aot_module_simplified(gm, example_inputs, **self.kwargs)
  File "c:\xinanlin\pytorch\torch\_functorch\aot_autograd.py", line 1168, in aot_module_simplified
    compiled_fn = dispatch_and_compile()
  File "c:\xinanlin\pytorch\torch\_functorch\aot_autograd.py", line 1143, in dispatch_and_compile
    compiled_fn, _ = create_aot_dispatcher_function(
  File "c:\xinanlin\pytorch\torch\_functorch\aot_autograd.py", line 570, in create_aot_dispatcher_function
    return _create_aot_dispatcher_function(
  File "c:\xinanlin\pytorch\torch\_functorch\aot_autograd.py", line 820, in _create_aot_dispatcher_function
    compiled_fn, fw_metadata = compiler_fn(
  File "c:\xinanlin\pytorch\torch\_functorch\_aot_autograd\jit_compile_runtime_wrappers.py", line 205, in aot_dispatch_base
    compiled_fw = compiler(fw_module, updated_flat_args)
  File "c:\xinanlin\pytorch\torch\_functorch\aot_autograd.py", line 479, in __call__
    return self.compiler_fn(gm, example_inputs)
  File "c:\xinanlin\pytorch\torch\_inductor\compile_fx.py", line 2043, in fw_compiler_base
    return inner_compile(
  File "c:\xinanlin\pytorch\torch\_inductor\compile_fx.py", line 623, in compile_fx_inner
    return wrap_compiler_debug(_compile_fx_inner, compiler_name="inductor")(
  File "c:\xinanlin\pytorch\torch\_dynamo\repro\after_aot.py", line 123, in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
  File "c:\xinanlin\pytorch\torch\_inductor\compile_fx.py", line 714, in _compile_fx_inner
    mb_compiled_graph, cache_info = FxGraphCache.load_with_key(
  File "c:\xinanlin\pytorch\torch\_inductor\codecache.py", line 1285, in load_with_key
    compiled_graph, cache_info = FxGraphCache._lookup_graph(
  File "c:\xinanlin\pytorch\torch\_inductor\codecache.py", line 1042, in _lookup_graph
    triton_bundler_meta = TritonBundler.read_and_emit(bundle)
  File "c:\xinanlin\pytorch\torch\_inductor\triton_bundler.py", line 268, in read_and_emit
    os.replace(tmp_dir, directory)
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
PermissionError: [WinError 5] Access is denied: 'C:\\Users\\sdp\\AppData\\Local\\Temp\\tmpayxttyg_\\triton\\tmp.5cbc7c41-fcc6-4f26-b2ea-8f244922901b' -> 'C:\\Users\\sdp\\AppData\\Local\\Temp\\tmpayxttyg_\\triton\\B2BWR5V2UZHVQTZBWRTHIDB66UFQM3AQIACWKYR434TZCG6VQYSA'

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```
I found that the `directory` is an **empty** directory created by the line:
https://github.com/pytorch/pytorch/blob/454fbd5bbe52d241a0ac89dc37574a16d322a1c0/torch/_inductor/triton_bundler.py#L241

It's not allowed to replace a director to an existing even **empty** director on Windows, following is my python test script:
```
(xinanlin) C:\xinanlin>ipython
Python 3.9.21 | packaged by conda-forge | (main, Dec  5 2024, 13:41:22) [MSC v.1929 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os

In [2]: os.mkdir("a")

In [3]: os.mkdir("b")

In [4]: os.replace("a", "b")
---------------------------------------------------------------------------
PermissionError                           Traceback (most recent call last)
Cell In[4], line 1
----> 1 os.replace("a", "b")

PermissionError: [WinError 5] Access is denied: 'a' -> 'b'
```

From the context, this line of code should not take effect because our target directory will ultimately be replaced.
So in my updated change, I removed this line and add the file lock for the `os.replace` for Windows.

etaf committed:
Update on "[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows."


This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

etaf committed:
Update on "[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows."


This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

oulgen commented on torch/_inductor/triton_bundler.py:
```diff
@@ -260,6 +259,11 @@ def read_and_emit(
                         kernel_names.append(Path(artifact.filename).stem)
-                # Atomic on POSIX systems
-                os.replace(tmp_dir, directory)
+                if _IS_WINDOWS:
```
did you test this? shouldnt this be not?

etaf replied:
> did you test this? shouldnt this be not?

Sorry for the mistake I made when I manually move the change from my windows to Linux.

EikanWang replied:
Does the code as follows make sense?
```python
file_lock = FileLock(directory + ".lock") if _IS_WINDOWS else contextlib.nullcontext()
with file_lock:
  os.replace(tmp_dir, directory)
```

etaf committed:
Update on "[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows."


This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

etaf committed:
Update on "[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows."


This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/triton_bundler.py:
```diff
@@ -260,6 +261,12 @@ def read_and_emit(
-                os.replace(tmp_dir, directory)
+
+                if _IS_WINDOWS:
+                    with FileLock(directory + ".lock"):
```
With the above basedir change is the filelock still needed?  Maybe we can just do that one.

etaf replied:
I think there may be a potential race condition with the target directory, as different threads might attempt to perform `os.replace` to the same target directory simultaneously.
@oulgen may I know your suggestion?

jansel replied:
But doesn't `os.replace` throw an error if it already exists on Windows? How does this fix the race? Or was I misreading the error message?

etaf replied:
Hi @jansel , the error message I met in https://github.com/pytorch/pytorch/pull/146481#discussion_r1961058510 is not about the race. It's ok to that the target directory of os.replace is existed on Linux but not for Windows. So I updated the line that creating the empty directory to creating it's parent directory (make sure `os.replace` works if the parent of target directory not created.). This makes sure that the creation/replacement of the target directory only happens on `os.replace`. So I add the filelock for the operation.

>But doesn't os.replace throw an error if it already exists on Windows?
Yes, I should add a line that remove the directory if exists on Windows.

etaf committed:
Update on "[Inductor UT][Windows][XPU] Fix Inductor UT on XPU Windows."


This PR fixed all the inductor UT failures for XPU backend on Windows we found in local machine(Due to resource constraints, we have not yet set up a Windows CI pipeline online.)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

etaf commented:
@pytorchbot merge
